workitem/link/link_repository.go

```diff
@@ -121,6 +108,98 @@ func (r *GormWorkItemLinkRepository) ValidateTopology(ctx context.Context, sourc
+// cycle when created.
+//
+// Legend
+// ------
+//
+//   \ = link
+//   * = new link
+//   C = the element that is potentially causing the cycle
+//
+// Scenarios
+// ---------
+//
+//   I:        II:       III:      IV:       V:       VI:
+//
+//    C         C         C         C         A        A
+//     *         \         *         *         \        \
+//      A         A         A         A         B        C
+//       \         \         \         \         *        \
+//        B         B         C         B         C        B
+//         \         *         \                            *
+//          C         C         B                            C
+//
+// In a "tree" topology:
+//   I, II, III are cycles
+//   IV and V are no cycles.
+//   VI violates the single-parent rule
+//
+// In a "dependency" topology:
+//   I, II, III, and VI are cycles
+//   IV and V are no cycles.
+//
+// Possibility to detect each cycle (if any)
+// -----------------------------------------
+//
+// In the existing topology we search for the new link's source and traverse up
+// to get its ancestors. If any of those ancestors match the new link's target,
+// we have found ourselves a cycle. Holds true for I, II, III, V, IV, and VI.
+func (r *GormWorkItemLinkRepository) DetectCycle(ctx context.Context, sourceID, targetID, linkTypeID uuid.UUID) (hasCycle bool, err error) {
+	// Get all roots for link's source.
+	// NOTE(kwk): Yes there can be more than one, if the link type is allowing it.
+	ancestors, err := r.GetAncestors(ctx, linkTypeID, sourceID)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"wilt_id":   linkTypeID,
+			"target_id": targetID,
+			"err":       err,
+		}, "failed to check if the work item %s has a parent work item", targetID)
+		return false, errs.Wrapf(err, "failed to check if the work item %s has a parent work item", targetID)
+	}
+
+	for _, ancestor := range ancestors {
+		if ancestor.ID == targetID { // Scenarios I, II, III, VI
+			return true, nil
+		}
+	}
+	return false, nil // Scenario IV and V
+}
+
+// acquireLock takes a space ID and acquires an advisory lock. It blocks until
+// it acquires the lock.
+func (r *GormWorkItemLinkRepository) acquireLock(spaceID uuid.UUID) error {
+	// hash the UUID to get a 32 bit value from a 128 UUID that we can then use
+	// with a PostgreSQL advisory lock.
+	//
+	// TODO(kwk): Reducing the 128 bit UUID to a 32 bit number potentially ends
+	// up in longer wait times for locks because there exists a certain
+	// potential to have hash collisions for two distinct UUIDs. But currently
+	// Postgres only allows for 64 bit numbers without the high bit set. That
+	// effectively limits us to just 32 bit hashes. Once we allow cross-space
+	// linking we need to revisit the locking.
+	h := fnv.New32()
+	h.Write([]byte(spaceID.String()))
+	key := h.Sum32()
+
+	db := r.db.Exec("SELECT pg_advisory_xact_lock($1)", key)
```

Comment:
just curious: how does the `SELECT pg_advisory_xact_lock($1)` statement knows on which table to obtain the lock ?

Reply:
Simple, the lock is not bound to a table ;)

Reply:
ohh... so there's no need to create a secondary table with all space IDs, as we discussed earlier? awesome!

Reply:
Correct. Just for the sake of it: I've prefixed the string input to the hash function with `"links-"` in a4a5232.